src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -555,16 +557,34 @@ protected synchronized int engineDoFinal(byte[] bytes, int offset, int length, b
             }
 
             if (opMode == NATIVE_MODE_DECRYPT) {
-                // We use this temporary buffer both for our input ciphertext and our output (possibly unauthenticated) plaintext.
-                // We'll copy it to the actual output location iff decryption completes successfully.
-                engineUpdate(bytes, offset, length); // Decrypt mode never generates output for updates
-                final byte[] tempBuffer = decryptInputBuf.getDataBuffer();
-                final int tempBufferLength = decryptInputBuf.size();
-
-                if (tempBufferLength < tagLength) {
+                // We cannot write directly to the output array because that could release
+                // possibly unauthenticated plaintext.
+                // If we have already processed some amount of ciphertext, then we can use our existing input buffer
+                // for both input and output and then copy it over to the real output upon success.
+                // If we have not already processed some ciphertext, then we can use the input buffer as is
+                // and allocate temporary output buffer.
+                if (decryptInputBuf.size() + length < tagLength) {
                     throw new AEADBadTagException("Input too short - need tag");
                 }
 
+                final byte[] workingInputArray;
+                final byte[] workingOutputArray;
+                final int workingInputOffset;
+                final int workingInputLength;
+                if (decryptInputBuf.size() > 0) {
+                    // We've already processed data so need to merge this in
+                    engineUpdate(bytes, offset, length); // Decrypt mode never generates output for updates
+                    workingInputArray = decryptInputBuf.getDataBuffer();
+                    workingInputLength = decryptInputBuf.size();
+                    workingInputOffset = 0;
+                    workingOutputArray = workingInputArray; // We can decrypt in place
```

Comment:
just to check my understanding -- this is where you're saving a copy, right? the idea is that instead of making [each call to EVP_CipherUpdate](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/csrc/aes_gcm.cpp#L81) do a bunch of extra loads/stores for some array elsewhere in memory, it's already got the (consolidated in/out) buffer loaded and ready to go?

Reply:
synced up offline -- my confusion was that I missed the [underlying copy](https://github.com/corretto/amazon-corretto-crypto-provider/blob/develop/src/com/amazon/corretto/crypto/provider/AesGcmSpi.java#L425) in the call to `engineUpdate` on line 576. in the one-shot case (i.e. the `else` branch below), we simply reference the input array and pass that down to AWS-LC instead of copying `bytes`'s contents into our instance buffer.